### PR TITLE
Support the null value in bloom_filter_agg Spark aggregate function

### PR DESCRIPTION
**Pull request by weixiuli**
Currently, the velox BloomFilterAggregate checks the input row and throws an exception if there are some null values in the row. So we need to be consistent with spark's behavior and ignore null values.

The spark  BloomFilterAggregate will Ignore null values. https://github.com/apache/spark/blob/6cdca10f148433664b3e2be6f655b0ddba817537/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/BloomFilterAggregate.scala#L180-L188 
```
 override def update(buffer: BloomFilter, inputRow: InternalRow): BloomFilter = {
    val value = child.eval(inputRow)
    // Ignore null values.
    if (value == null) {
      return buffer
    }
    updater.update(buffer, value)
    buffer
  }
```

The unittest:
```scala
class VeloxBloomFilteAggregateFunctionsSuite extends VeloxWholeStageTransformerSuite {
  val funcId_bloom_filter_agg = new FunctionIdentifier("bloom_filter_agg")
  override protected val backend: String = "velox"
  override protected val resourcePath: String = "/tpch-data-parquet-velox"
  override protected val fileFormat: String = "parquet"
  val table = "bloomTable"

  protected def registerFunAndcCreatTable(): Unit ={
    val funcId_bloom_filter_agg = new FunctionIdentifier("bloom_filter_agg")
    // Register 'bloom_filter_agg'
    spark.sessionState.functionRegistry.registerFunction(funcId_bloom_filter_agg,
      new ExpressionInfo(classOf[BloomFilterAggregate].getName, "bloom_filter_agg"),
      (children: Seq[Expression]) => children.size match {
        case 1 => new BloomFilterAggregate(children.head)
        case 2 => new BloomFilterAggregate(children.head, children(1))
        case 3 => new BloomFilterAggregate(children.head, children(1), children(2))
      })
    val schema2 = new StructType().add("a2", IntegerType, nullable = true)
      .add("b2", LongType, nullable = true)
      .add("c2", IntegerType, nullable = true)
      .add("d2", IntegerType, nullable = true)
      .add("e2", IntegerType, nullable = true)
      .add("f2", IntegerType, nullable = true)
    val data2 = Seq(Seq(67, 17L, 45, 91, null, null),
      Seq(98, 63L, 0, 89, null, 40),
      Seq(null, null, 68, 75, 20, 19))
    val rdd2 = spark.sparkContext.parallelize(data2)
    val rddRow2 = rdd2.map(s => Row.fromSeq(s))
    spark.createDataFrame(rddRow2, schema2).write.saveAsTable(table)
  }

  protected def dropFunctionAndTable(): Unit ={
    spark.sessionState.functionRegistry.dropFunction(funcId_bloom_filter_agg)
    spark.sql(s"DROP TABLE IF EXISTS $table")
  }

  override def beforeAll(): Unit = {
    super.beforeAll()
    registerFunAndcCreatTable()
  }
  override def afterAll(): Unit = {
    dropFunctionAndTable()
    super.afterAll()
  }

  test("Test bloom_filter_agg with Nulls input") {
    spark.sql(
      s"""
       SELECT  bloom_filter_agg(b2) from $table
         """.stripMargin).show
  }
```

### CODE REVIEW

**Comment by weixiuli** [comment]
ping @marin-ma @jinchengchenghh Kindly reivew, thanks.

**Comment by jinchengchenghh** [comment]
The bloom_filter_agg is opensource Spark internal function, all the first argument of it is wrapped by xxhash, so it won't be null.
But for bloom_filter_agg itself, it can accept null value as this PR shows.
This case may occur because they use it's internal Spark version which change the usage of the function.
We could accept this PR because it aligns with Spark function behavior. 
What do you think? @rui-mo 
image

**Comment by jinchengchenghh** [comment]
Please also update the document https://github.com/facebookincubator/velox/blob/main/velox/docs/functions/spark/aggregate.rst?plain=1#L27

File: velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp [hunk 1]
```diff
@@ -100,10 +100,10 @@ class BloomFilterAggAggregate : public exec::Aggregate {
       bool /*mayPushdown*/) override {
     decodeArguments(rows, args);
     computeCapacity();
-    auto mayHaveNulls = decodedRaw_.mayHaveNulls();
+
     rows.applyToSelected([&](vector_size_t row) {
-      if (mayHaveNulls) {
-        checkBloomFilterNotNull(decodedRaw_, row);
+      if (decodedRaw_.isNullAt(row)) {
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp:105]: Check the `mayHaveNulls` first, and then check if this row is null.
Most cases `mayHaveNulls` is false, it will be more efficient.

File: velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp [hunk 2]
```diff
@@ -100,10 +100,10 @@ class BloomFilterAggAggregate : public exec::Aggregate {
       bool /*mayPushdown*/) override {
     decodeArguments(rows, args);
     computeCapacity();
-    auto mayHaveNulls = decodedRaw_.mayHaveNulls();
+
     rows.applyToSelected([&](vector_size_t row) {
-      if (mayHaveNulls) {
-        checkBloomFilterNotNull(decodedRaw_, row);
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp:106]: Please delete the function `checkBloomFilterNotNull`

**Comment by jinchengchenghh** [comment]
Please update the title to `Support the null values in bloom_filter_agg Spark aggregate function`

File: velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp [hunk 3]
```diff
@@ -144,17 +144,17 @@ class BloomFilterAggAggregate : public exec::Aggregate {
...
+      if (!decodedRaw_.isNullAt(0)) {
+        accumulator->insert(decodedRaw_.valueAt<int64_t>(0));
       }
-      accumulator->insert(decodedRaw_.valueAt<int64_t>(row));
-    });
+    } else {
+      rows.applyToSelected([&](vector_size_t row) {
+        if (decodedRaw_.isNullAt(row)) {
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp:152]: same

**Comment by weixiuli** [comment]
@jinchengchenghh @@rui-mo Kindly reivew, thanks.

File: velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp [hunk 4]
```diff
@@ -144,17 +136,18 @@ class BloomFilterAggAggregate : public exec::Aggregate {
...
-    rows.applyToSelected([&](vector_size_t row) {
-      if (mayHaveNulls) {
-        checkBloomFilterNotNull(decodedRaw_, row);
+      if (!decodedRaw_.isNullAt(0)) {
+        accumulator->insert(decodedRaw_.valueAt<int64_t>(0));
       }
-      accumulator->insert(decodedRaw_.valueAt<int64_t>(row));
-    });
+    } else {
+      auto mayHaveNulls = decodedRaw_.mayHaveNulls();
```
Comment (rui-mo) [inline velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp:143]: Seems we don't need to check `mayHaveNulls` for each input. Could be separated as below.
```
  const auto mayHaveNulls = decodedRaw_.mayHaveNulls();
  if UNLIKELY (mayHaveNulls) {
    rows.applyToSelected([&](vector_size_t row) {
      if UNLIKELY (decodedRaw_.isNullAt(row)) {
        return;
      }
      accumulator->insert(decodedRaw_.valueAt<int64_t>(row));
    }
    return;
  }
  rows.applyToSelected([&](vector_size_t row) {
    accumulator->insert(decodedRaw_.valueAt<int64_t>(row));
  }
```

Comment (rui-mo) [inline velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp:143]: const

File: velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp [hunk 5]
```diff
@@ -102,8 +94,8 @@ class BloomFilterAggAggregate : public exec::Aggregate {
     computeCapacity();
     auto mayHaveNulls = decodedRaw_.mayHaveNulls();
     rows.applyToSelected([&](vector_size_t row) {
-      if (mayHaveNulls) {
-        checkBloomFilterNotNull(decodedRaw_, row);
+      if (mayHaveNulls && decodedRaw_.isNullAt(row)) {
```
Comment (rui-mo) [inline velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp:97]: Could be optimized similarly by separating into different branches. Maybe wrap the common part in a lambda to avoid copy-paste.

**Comment by rui-mo** [comment]
@jinchengchenghh Thanks for your comment. It looks good to align with Spark's behavior.

**Comment by weixiuli** [comment]
@rui-mo @jinchengchenghh  PTAL.

File: velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp [hunk 6]
```diff
@@ -85,32 +85,32 @@ class BloomFilterAggAggregate : public exec::Aggregate {
...
+          return;
+        }
+        insert(groups, row);
+      });
+      return;
+    }
+    rows.applyToSelected([&](vector_size_t row) { insert(groups, row); });
+  }
+
+  void insert(char** groups, vector_size_t row) {
```
Comment (rui-mo) [inline velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp:108]: This method could be private.

**Review by rui-mo — COMMENTED** [review]
Thanks for working on this.

**Comment by weixiuli** [comment]
@rui-mo @jinchengchenghh PTAL.

**Comment by rui-mo** [comment]
@weixiuli Thanks for iterating. I have no further comment.

**Comment by jinchengchenghh** [comment]
> Please also update the document https://github.com/facebookincubator/velox/blob/main/velox/docs/functions/spark/aggregate.rst?plain=1#L27

please note this comment

**Comment by weixiuli** [comment]
@mbasmanova Kindly reivew, thanks.

**Comment by mbasmanova** [comment]
@weixiuli I googled for bloom_filter_agg Spark function, but nothing came up. Would you share a link to Spark documentation for this function? I'm trying to understand what has changed from the time this function was added to Velox and now. Was the original implementation incorrect? How can we ensure that this implementation is right?

**Comment by weixiuli** [comment]
> @weixiuli I googled for bloom_filter_agg Spark function, but nothing came up. Would you share a link to Spark documentation for this function? I'm trying to understand what has changed from the time this function was added to Velox and now. Was the original implementation incorrect? How can we ensure that this implementation is right?

Thanks @mbasmanova  I've already described the differences between Velox and spark in BloomFilterAggregate at the beginning of PR, please take a look and the  `the link to Spark documentation for this function`  is 
https://github.com/apache/spark/blob/6cdca10f148433664b3e2be6f655b0ddba817537/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/BloomFilterAggregate.scala#L180-L188

**Comment by weixiuli** [comment]
> > @weixiuli I googled for bloom_filter_agg Spark function, but nothing came up. Would you share a link to Spark documentation for this function? I'm trying to understand what has changed from the time this function was added to Velox and now. Was the original implementation incorrect? How can we ensure that this implementation is right?
> 
> Thanks @mbasmanova I've already described the differences between Velox and spark in BloomFilterAggregate at the beginning of PR, please take a look and the `the link to Spark documentation for this function` is https://github.com/apache/spark/blob/6cdca10f148433664b3e2be6f655b0ddba817537/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/BloomFilterAggregate.scala#L180-L188

@mbasmanova PTAL.

**Comment by mbasmanova** [comment]
@weixiuli 

> Would you share a link to Spark documentation for this function? I'm trying to understand what has changed from the time this function was added to Velox and now. Was the original implementation incorrect? How can we ensure that this implementation is right?

Would you answer all of these questions? The link you provided points to some code in Spark. Code is not documentation.

File: velox/docs/functions/spark/aggregate.rst [hunk 7]
```diff
@@ -24,7 +24,6 @@ General Aggregate Functions
 .. spark:function:: bloom_filter_agg(hash, estimatedNumItems, numBits) -> varbinary
 
     Creates bloom filter from input hashes and returns it serialized into VARBINARY.
-    The caller is expected to apply xxhash64 function to input data before calling bloom_filter_agg.
```
Comment (mbasmanova) [inline velox/docs/functions/spark/aggregate.rst:27]: Why did you remove this sentence? Is it incorrect?

At the same time, I see that "``hash`` cannot be null." is still part of the docs. Please, take time to carefully review the documentation and make sure it is accurate and complete.

File: velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp [hunk 8]
```diff
@@ -71,14 +71,12 @@ TEST_F(BloomFilterAggAggregateTest, emptyInput) {
...
-TEST_F(BloomFilterAggAggregateTest, nullBloomFilter) {
-  auto vectors = {makeRowVector({makeAllNullFlatVector<int64_t>(2)})};
-  auto expectedFake = {makeRowVector(
-      {makeNullableFlatVector<StringView>({std::nullopt}, VARBINARY())})};
-  VELOX_ASSERT_THROW(
-      testAggregations(
-          vectors, {}, {"bloom_filter_agg(c0, 5, 64)"}, expectedFake),
-      "First argument of bloom_filter_agg cannot be null");
+TEST_F(BloomFilterAggAggregateTest, nullInput) {
+  auto vectors = {makeRowVector(
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp:75]: It seems like we create 2 columns here: c0 and c1. c0 has no nulls. c1 has only nulls. However, testAggregations only uses c0. Why do we need c1? Does this test actually verify the changes? Would this test fail without the code changes? 

Also, notice that sizes of columns c0 and c1 and different. c0 has 100 rows, while c1 only 1. 

Please, take the time to carefully review the test and make sure it actually verifies the changes and doesn't include unnecessary code.

File: velox/docs/functions/spark/aggregate.rst (hunk 7 above)
Comment (weixiuli) [reply velox/docs/functions/spark/aggregate.rst:27]: The `bloom_filter_agg` can support `null` values and is no longer expected to apply the `xxhash64` function to input data before calling `bloom_filter_agg`. I have updated the docs.

File: velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp (hunk 8 above)
Comment (weixiuli) [reply velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp]: Yes. I have updated the unittest.

**Comment by weixiuli** [comment]
> Would you answer all of these questions? The link you provided points to some code in Spark. Code is not documentation.

The Spark BloomFilterAggregate document can refer to https://books.japila.pl/spark-sql-internals/expressions/BloomFilterAggregate/

**Comment by weixiuli** [comment]
@mbasmanova Kindly review, thanks.

File: velox/docs/functions/spark/aggregate.rst [hunk 9]
```diff
@@ -24,14 +24,13 @@ General Aggregate Functions
...
 
     For example, 
         bloom_filter_agg(xxhash64(x), 100, 1024)
     In Spark implementation, ``estimatedNumItems`` and ``numBits`` are used to decide the number of hash functions and bloom filter capacity.
     In Velox implementation, ``estimatedNumItems`` is not used.
 
-    ``hash`` cannot be null.
+    ``hash`` can be null.
```
Comment (mbasmanova) [inline velox/docs/functions/spark/aggregate.rst:33]: L44 needs updating as well. Please, do review all the documentation to make sure it is updated to accommodate changes.

File: velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp [hunk 10]
```diff
@@ -71,14 +75,12 @@ TEST_F(BloomFilterAggAggregateTest, emptyInput) {
...
-  auto expectedFake = {makeRowVector(
-      {makeNullableFlatVector<StringView>({std::nullopt}, VARBINARY())})};
-  VELOX_ASSERT_THROW(
-      testAggregations(
-          vectors, {}, {"bloom_filter_agg(c0, 5, 64)"}, expectedFake),
-      "First argument of bloom_filter_agg cannot be null");
+TEST_F(BloomFilterAggAggregateTest, nullInput) {
+  auto vectors = {makeRowVector({makeAllNullFlatVector<int64_t>(1)})};
+  // The bloom filter supports the input data are nulls and will skip them.
+  // The expected value is the serialized bloom filter without any input values.
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp:81]: This is surprising. I'd expect null input to produce null result. If that's not the case then let's update documentation to clarify the result for null input.

File: velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp [hunk 11]
```diff
@@ -71,14 +75,12 @@ TEST_F(BloomFilterAggAggregateTest, emptyInput) {
...
-TEST_F(BloomFilterAggAggregateTest, nullBloomFilter) {
-  auto vectors = {makeRowVector({makeAllNullFlatVector<int64_t>(2)})};
-  auto expectedFake = {makeRowVector(
-      {makeNullableFlatVector<StringView>({std::nullopt}, VARBINARY())})};
-  VELOX_ASSERT_THROW(
-      testAggregations(
-          vectors, {}, {"bloom_filter_agg(c0, 5, 64)"}, expectedFake),
-      "First argument of bloom_filter_agg cannot be null");
+TEST_F(BloomFilterAggAggregateTest, nullInput) {
+  auto vectors = {makeRowVector({makeAllNullFlatVector<int64_t>(1)})};
```
[the remark this answers is not in the document]
Comment (weixiuli) [reply velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp]: Done.

**Comment by weixiuli** [comment]
@mbasmanova  Could you help me review this pr again, thanks.

**Comment by mbasmanova** [comment]
@weixiuli I was out sick last week. I'm back today, but still recovering. Would you rebase this PR? I'll take a look sometime this week.

File: velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp [hunk 12]
```diff
@@ -71,14 +71,15 @@ TEST_F(BloomFilterAggAggregateTest, emptyInput) {
...
-  auto vectors = {makeRowVector({makeAllNullFlatVector<int64_t>(2)})};
-  auto expectedFake = {makeRowVector(
-      {makeNullableFlatVector<StringView>({std::nullopt}, VARBINARY())})};
-  VELOX_ASSERT_THROW(
-      testAggregations(
-          vectors, {}, {"bloom_filter_agg(c0, 5, 64)"}, expectedFake),
-      "First argument of bloom_filter_agg cannot be null");
+TEST_F(BloomFilterAggAggregateTest, withNullInput) {
+  // The first 10 values range from 0 to 9, and the last 10 values are nulls
+  auto vectors = {makeRowVector({makeFlatVector<int64_t>(
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp:76]: This can be written in a more reader-friendly way.

```
makeNullableFlatVector<int64_>({0, 1, 2, std::nullopt, 3, 4, 5, std::nullopt, std::nullopt, 6, 7, 8, 9, std::nullopt, std::nullopt, std::nullopt});
```

File: velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp [hunk 13]
```diff
@@ -71,14 +71,15 @@ TEST_F(BloomFilterAggAggregateTest, emptyInput) {
...
+      20,
+      [](vector_size_t row) { return row % 9; },
+      [](vector_size_t row) { return row > 9; })})};
+  // The expected value is the serialized bloom filter
+  auto expected = {makeRowVector({getSerializedBloomFilter(4)})};
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp:81]: Can we also test all-null input? I remember seeing that it had a somewhat unexpected (non-NULL) result. Let's document what that result is.

**Comment by weixiuli** [comment]
@mbasmanova PTAL.

File: velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp [hunk 14]
```diff
@@ -71,14 +75,47 @@ TEST_F(BloomFilterAggAggregateTest, emptyInput) {
...
+       std::nullopt,
+       std::nullopt})})};
+  // The expected value is the serialized bloom filter
+  auto expected = {makeRowVector({getSerializedBloomFilter(4)})};
+  testAggregations(vectors, {}, {"bloom_filter_agg(c0, 5, 64)"}, expected);
+}
+
+TEST_F(BloomFilterAggAggregateTest, allNullInput) {
+  // The bloom_filter_agg do not support test streaming when it's inputs are all
+  // NULL, because the result of streaming will be null
```
[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp:101]: cc: @jinchengchenghh

[the remark this answers is not in the document]
Comment (jinchengchenghh) [reply velox/functions/sparksql/aggregates/tests/BloomFilterAggAggregateTest.cpp:101]: As this comment indicates https://github.com/facebookincubator/velox/pull/7872#issuecomment-1853166379, all null input will be xxhashed in opensource and then input to bloom_filter_agg, so this function of Spark will not process all null input. What's the error in Velox? @weixiuli
